pkg/query-service/app/traces/v4/query_builder_test.go

```diff
@@ -0,0 +1,556 @@
+				"timestamp <= '1680066458000000000') AND (ts_bucket_start >= 1680064560 AND ts_bucket_start <= 1680066458) GROUP BY subQuery.traceID, subQuery.durationNano, subQuery.name, subQuery.serviceName ORDER BY " +
+				"subQuery.durationNano desc LIMIT 1 BY subQuery.traceID;",
+		},
+		{
+			name: "Test order by value with having",
+			args: args{
+				panelType: v3.PanelTypeTable,
+				start:     1680066360726210000,
+				end:       1680066458000000000,
+				mq: &v3.BuilderQuery{
+					AggregateOperator:  v3.AggregateOperatorCountDistinct,
+					Filters:            &v3.FilterSet{},
+					AggregateAttribute: v3.AttributeKey{Key: "name", IsColumn: true, DataType: v3.AttributeKeyDataTypeString, Type: v3.AttributeKeyTypeTag},
+					OrderBy:            []v3.OrderBy{{ColumnName: "#SIGNOZ_VALUE", Order: "ASC"}},
+					Having: []v3.Having{
+						{
+							ColumnName: "name",
+							Operator:   ">",
+							Value:      10,
+						},
+					},
+				},
+			},
+			want: "SELECT  toFloat64(count(distinct(name))) as value from signoz_traces.distributed_signoz_index_v3 where (timestamp >= '1680066360726210000' AND timestamp <= '1680066458000000000') AND " +
+				"(ts_bucket_start >= 1680064560 AND ts_bucket_start <= 1680066458) having value > 10 order by value ASC",
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := buildTracesQuery(tt.args.start, tt.args.end, tt.args.step, tt.args.mq, tt.args.panelType, tt.args.options)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("buildTracesQuery() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if got != tt.want {
+				t.Errorf("buildTracesQuery() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func TestPrepareTracesQuery(t *testing.T) {
+	type args struct {
+		start     int64
+		end       int64
+		panelType v3.PanelType
+		mq        *v3.BuilderQuery
+		options   v3.QBOptions
+	}
+	tests := []struct {
+		name    string
+		args    args
+		want    string
+		wantErr bool
+	}{
+		{
+			name: "test with limit - first",
+			args: args{
+				start:     1680066360726210000,
+				end:       1680066458000000000,
+				panelType: v3.PanelTypeTable,
+				mq: &v3.BuilderQuery{
+					StepInterval:       60,
+					AggregateOperator:  v3.AggregateOperatorCountDistinct,
+					Filters:            &v3.FilterSet{},
+					AggregateAttribute: v3.AttributeKey{Key: "name", IsColumn: true, DataType: v3.AttributeKeyDataTypeString, Type: v3.AttributeKeyTypeTag},
+					GroupBy:            []v3.AttributeKey{{Key: "serviceName", IsColumn: true, DataType: v3.AttributeKeyDataTypeString, Type: v3.AttributeKeyTypeTag}},
+					Limit:              10,
+				},
+				options: v3.QBOptions{
+					GraphLimitQtype: constants.FirstQueryGraphLimit,
+				},
+			},
+			want: "SELECT `serviceName` from (SELECT serviceName as `serviceName`, toFloat64(count(distinct(name))) as value from signoz_traces.distributed_signoz_index_v3 " +
```

Comment:
There is no order by for value, would that mean the result can be the random and inconsistent?
```sql
SELECT `serviceName` 
FROM (
    SELECT 
        serviceName as `serviceName`,
        toFloat64(count(distinct(name))) as value 
    FROM signoz_traces.distributed_signoz_index_v3 
    WHERE (
        timestamp >= '1680066360726210000' 
        AND timestamp <= '1680066458000000000'
    ) 
    AND (
        ts_bucket_start >= 1680064560 
        AND ts_bucket_start <= 1680066458
    ) 
    GROUP BY `serviceName`
) 
LIMIT 10
```

Reply:
Have added order by in the query. The test was invalid, thanks for pointing out.

Reply:
What happens if there is no order by present in the request payload?